packages/preact-iso/router.js

```diff
@@ -1,16 +1,34 @@
 import { h, createContext, cloneElement } from 'preact';
 import { useContext, useMemo, useReducer, useEffect, useLayoutEffect, useRef } from 'preact/hooks';
 
-const UPDATE = (state, url, push) => {
-	if (url && url.type === 'click') {
+/**
+ * @param {string} state
+ * @param {MouseEvent|PopStateEvent|Object|string} url
+ * @return {string}
+ */
+const UPDATE = (state, url) => {
+	/** @type {boolean|undefined} - History state update strategy */
+	let push = undefined
+
+	// user click
+	if (url instanceof MouseEvent) {
+		// @ts-ignore-next
 		const link = url.target.closest('a[href]');
 		if (!link || link.origin != location.origin) return state;
 
 		url.preventDefault();
 		push = true;
 		url = link.href.replace(location.origin, '');
-	} else if (typeof url !== 'string') {
+	// navigation
+	} else if (url instanceof PopStateEvent) {
 		url = location.pathname + location.search;
+	// manual invocation: route({ path, replace })
+	} else if (typeof url === 'object') {
```

Comment:
Here's what I'd suggest, in order to preserve the duck typing (as mentioned in previous comment):

```js
const UPDATE = (state, update) => {
	/** @type {boolean|undefined} - History state update strategy */
	let push, url;

	if (!update || typeof update === 'string') {
		url = update;
		push = true;
	}
	else if (update.type === 'click') {
		// user click
		// @ts-ignore-next
		const link = update.target.closest('a[href]');
		if (!link || link.origin != location.origin) return state;

		update.preventDefault();
		url = link.pathname + link.search + link.hash;
		push = true;
	}
	else if (update.type === 'popstate') {
		// navigation
		url = location.pathname + location.search + location.hash;
	}
	else {
		// manual invocation: route(url) or route({ url, replace })
		url = update.url || update;
		push = !url.replace;
	}

	if (push === true) history.pushState(null, '', url);
	else if (push === false) history.replaceState(null, '', url);
	return url;
}
```

Reply:
Should we allow passing a value that evaluates to false (undefined/ null) with route method?
If so, then should it be added to history?
```js
	if (!update || typeof update === 'string') {
	//  ^
```

Are if the `update` is not string or event I'd say it must be an object?
```js
	else {
		// manual invocation: route(url) or route({ url, replace })
		url = update.url || update;
		//                  ^
```

Reply:
I've updated PR with code you suggested, adapting to eslint/ prettier rules

Reply:
@developit 
What do you think about my comments above?